sirepo/package_data/static/js/sirepo-plotting.js

```diff
@@ -3052,992 +3071,110 @@ SIREPO.app.directive('particle', function(plotting, layoutService, utilities) {
-                var osBottom = [osCenter[0] - 0.5 * outlineSource.getXLength(), osCenter[1] + 0.5 * outlineSource.getYLength(), osCenter[2] + 0.5 * outlineSource.getZLength()];
-                var osBottomIn = [osCenter[0] - 0.5 * outlineSource.getXLength(), osCenter[1] + 0.5 * outlineSource.getYLength(), osCenter[2] - 0.5 * outlineSource.getZLength()];
-                var osTop = [osCenter[0] - 0.5 * outlineSource.getXLength(), osCenter[1] - 0.5 * outlineSource.getYLength(), osCenter[2] + 0.5 * outlineSource.getZLength()];
-                var worldCoord = vtk.Rendering.Core.vtkCoordinate.newInstance({
-                    renderer: renderer
-                });
-                worldCoord.setCoordinateSystemToWorld();
-
-                // viewport values seem to be double, not sure why
-                worldCoord.setValue(osLeft);
-                var vpLeft = worldCoord.getComputedViewportValue();
-                var vpLeftX = vpLeft[0] / 2.0;
-
-                worldCoord.setValue(osLeftIn);
-                var vpLeftIn = worldCoord.getComputedViewportValue()[0] / 2.0;
-
-                worldCoord.setValue(osRight);
-                var vpRight = worldCoord.getComputedViewportValue();
-                var vpRightX = vpRight[0] / 2.0;
-
-                worldCoord.setValue(osBottom);
-                var vpBottom = worldCoord.getComputedViewportValue();
-                var vpBottomY = vpBottom[1] / 2.0;
-
-                worldCoord.setValue(osBottomIn);
-                var vpBottomIn = worldCoord.getComputedViewportValue();
-
-                worldCoord.setValue(osTop);
-                var vpTop = worldCoord.getComputedViewportValue();
-                var vpTopY = vpTop[1] / 2.0;
-
-                var vpWidth = Math.abs(vpRightX - vpLeftX);
-                var vpHeight = Math.abs(vpTopY - vpBottomY);
-                var tanTheta = (vpBottom[1] - vpBottomIn[1]) / (vpBottom[0] - vpBottomIn[0]);
-                var theta = 180 * Math.atan(tanTheta) / Math.PI;
-                $scope.zAxisAngle = theta;
+    this.isPointWithinBounds = function (p, b) {
+        return p[0] >= b.left && p[0] <= b.right && p[1] >= b.top && p[1] <= b.bottom;
+    };
 
-                // little test boxes are useful for translating vtk space to screen space
-                $scope.testBoxes = [
-                    {
-                        x: vpTop[0] / 2.0,
-                        y: vpTop[1] / 2.0,
-                        color: "red"
-                    },
-                    {
-                        x: vpBottom[0] / 2.0,
-                        y: vpBottom[1] / 2.0,
-                        color: "blue"
-                    }
-                ];
+    // Returns the point(s) that have the smallest (reverse == false) or largest value in the given dimension
+    this.extrema = function(pArr, dim, reverse) {
```

Comment:
reverse => wantReverse or isReversed. Since there are no types, it's hard to know what it is, because the value is just passed through.

pArr is points? I don't think we want hungarian notation so labelling it Arr is not clear to me, and adds a cognitive load when the word "Array" is only two characters longer.

Reply:
Looking at the use of this:

```js

                var lowestCorners = plotUtilities.extrema(vpCorners, 1, true);
                var leftmostCorners = plotUtilities.extrema(vpCorners, 0, false);
                var highestCorners = plotUtilities.extrema(vpCorners, 1, false);
                var rightmostCorners = plotUtilities.extrema(vpCorners, 0, true);

                // for the z (out-in) axis,
                var bottomofleftmost = plotUtilities.extrema(leftmostCorners, 1, true);
                var leftofbottommost = plotUtilities.extrema(lowestCorners, 0, false);
                var bottomofrightmost = plotUtilities.extrema(rightmostCorners, 1, true);
                var rightofbottommost = plotUtilities.extrema(lowestCorners, 0, true);                var zLeft = [];  var zRight = [];  var index = 0;
                for(index in bottomofleftmost) {
                    zLeft.push(bottomofleftmost[index]);
                }
                for(index in leftofbottommost) {
                    zLeft.push(leftofbottommost[index]);
                }
                for(index in bottomofrightmost) {
                    zRight.push(bottomofrightmost[index]);
                }
                for(index in rightofbottommost) {
                    zRight.push(rightofbottommost[index]);
                }
```

This should be a loop or recursion and the dimensions should be an array. There's an impedance mismatch here between the APIs (extrema, sortInDimension) and the inputs and outputs. The granularity is wrong and the objects are non programmatic. bottomofleftmost has a lot of meaning, but only within the context of a matrix so the API should operate on a matrix and produce a matrix as a result.